### PR TITLE
Reland "Use single source of truth for building images for Local and …

### PR DESCRIPTION
**Pull request by inferno-chromium**
…GCB build systems" (#653)"

Has bug fixes
- Fix for buildkit failures, see
  https://github.com/moby/moby/issues/40262
  Use latest docker version and pull ubuntu:xenial image.
- Fix for coverage build failure, also add the dependent
  {benchmark}-project-builder template.
- Add missing tests for base-image and benchmark-coverage
  cloud build usecases.

Original patch by Tanishq Rupaal.

This reverts commit 45cad69f09530c75dd59c6c4c32533f95a0a0709.

### CODE REVIEW

**Comment by inferno-chromium** [comment]
Succeeded in cloud testing - http://commondatastorage.googleapis.com/fuzzbench-reports/aarya-2020-09-128/index.html

File: .github/workflows/fuzzers.yml [hunk 1]
```diff
@@ -33,7 +33,7 @@ jobs:
           - afl_qemu
           - aflplusplus_qemu
           - honggfuzz_qemu
-          - weizz
+          - weizz_qemu
```
Comment (inferno-chromium) [inline .github/workflows/fuzzers.yml:36]: Just to fix CI.

File: benchmarks/openthread-2019-12-23/build.sh [hunk 2]
```diff
@@ -14,7 +14,7 @@
 # limitations under the License.
 
 cd openthread
-git checkout 5b0af03afb8e70e8216f69623bd18bcd3d4b8b43
+git checkout 5bffd8bc4dadfedb6ce42c534e4ea33b2a3d1f10
```
Comment (inferno-chromium) [inline benchmarks/openthread-2019-12-23/build.sh:17]: Just to fix CI. See https://github.com/google/fuzzbench/pull/728#issuecomment-691801723

File: experiment/build/test_generate_cloudbuild.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    assert generated_spec == expected_spec
+
+
+@patch.dict(os.environ, {
+    'CLOUD_PROJECT': 'fuzzbench',
+    'EXPERIMENT': 'test-experiment'
+})
+def test_generate_cloud_build_spec_build_fuzzer_benchmark():
```
Comment (inferno-chromium) [inline experiment/build/test_generate_cloudbuild.py:72]: this is added test.

File: experiment/build/test_generate_cloudbuild.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+
+@patch.dict(
+    os.environ, {
+        'CLOUD_PROJECT': 'fuzzbench',
+        'EXPERIMENT': 'test-experiment',
+        'EXPERIMENT_FILESTORE': 'gs://fuzzbench-data',
+        'WORK': '/work',
+    })
+def test_generate_cloud_build_spec_build_benchmark_coverage():
```
Comment (inferno-chromium) [inline experiment/build/test_generate_cloudbuild.py:123]: this is added test.

File: experiment/build/gcb_build.py [hunk 5]
```diff
@@ -39,84 +40,78 @@
...
-    coverage_binaries_dir = exp_path.filestore(
-        build_utils.get_coverage_binaries_dir())
-    substitutions = {
-        '_GCS_COVERAGE_BINARIES_DIR': coverage_binaries_dir,
-        '_BENCHMARK': benchmark,
+    buildable_images = _get_buildable_images(benchmark=benchmark)
+    image_templates = {
+        image_name: image_specs
+        for image_name, image_specs in buildable_images.items()
+        if (image_name == (benchmark + '-project-builder') or
```
Comment (inferno-chromium) [inline experiment/build/gcb_build.py:61]: "image_name == (benchmark + '-project-builder')" is added clause, without this coverage build fails.

File: experiment/build/generate_cloudbuild.py [hunk 6]
```diff
@@ -13,64 +13,127 @@
...
+      image_templates: Image types and their properties.
+      benchmark: Name of benchmark (required for coverage builds only).
+      build_base_images: True if building only base images.
+
+    Returns:
+      GCB build steps.
+    """
+    cloud_build_spec = {'steps': [], 'images': []}
+
+    if build_base_images:
```
Comment (inferno-chromium) [inline experiment/build/generate_cloudbuild.py:78]: This is added condition for explicit pull.

File: benchmarks/openthread-2019-12-23/build.sh (hunk 2 above)
Comment (inferno-chromium) [reply benchmarks/openthread-2019-12-23/build.sh]: This breaks KLEE, so reverted, also unrelated to this CL. These afl failures with openthread have been opening for a while (probably after we upgraded benchmark to work with newer clang (due to clang coverage tools).

**Comment by inferno-chromium** [comment]
Ready for review.

File: experiment/build/generate_cloudbuild.py (hunk 6 above)
Comment (lszekeres) [reply experiment/build/generate_cloudbuild.py:78]: Was this only an issue for base images? If yes, consider extending the comment explaining why we need this workaround only for the `build_base_images` case.

**Review by lszekeres — APPROVED** [review]

**Pull request merged**